### PR TITLE
[OCPBUGS-3341] Add RHOSP CCM options reference documentation

### PR DESCRIPTION
**Pull request by maxwelldb**
Version(s):
4.12

Issue:
[OCPBUGS-3341](https://issues.redhat.com/browse/OCPBUGS-3341)

Link to docs preview: https://53068--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-cloud-config-reference.html

QE review:
- [x] QE has approved this change.

Additional information: Need to figure out where to put this. It would bloat the installation assemblies. Possible options:

* Collapsible tables in installation assemblies. 
* **Dedicated RHOSP installation reference assembly**
* ???

Will link from installation assemblies to this one after this is merged.

### CODE REVIEW

**Comment by maxwelldb** [comment]
@MaysaMacedo @rlobillo This is brand new and not yet situated in the docs, but the module should be ready for a first look if you have time. 

Related to: https://github.com/openshift/openshift-docs/pull/52294

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+// | ID of the Neutron network on which to create the members of the load balancer. The load balancer gets another network port on this subnet. Defaults to `subnet-id` if not set.
+
+| `network-id`
+| The ID of the Networking service network on which load balancer VIPs are created. Unnecessary if `subnet-id` is set.
+
+// | `manage-security-groups`
+// | If the Neutron security groups should be managed separately. Default: false
+
+| `create-monitor`
+| Whether or not to create a health monitor for the service load balancer. A health monitor is required for services that declare `externalTrafficPolicy: Local`. The default value is `false`.
```
Comment (MaysaMacedo) [inline modules/cluster-cloud-controller-config-osp.adoc:193]: perhaps we should mention that when using OVN driver monitor is only supported from OSP17

**Comment by MaysaMacedo** [comment]
Should this same doc be included since 4.9?
There are certain options that only are being tested in CCM on 4.12

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+// | `container-store`
+// | Optional. Used to specify the store of the tls-container-ref, e.g. "barbican" or "external" - other store will cause a warning log. Default value - `barbican` - existence of tls container ref would always be performed. If set to `external` format for tls container ref will not be validated.
+
+| `max-shared-lb`
+| The maximum number of Services that can share a load balancer. The default value is `2`.
+|===
+
+[id="ccm-config-metadata-options"]
+== Metadata options
```
Comment (MaysaMacedo) [inline modules/cluster-cloud-controller-config-osp.adoc:257]: I don't think these options are tested

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+. The directory `~/.config/openstack`.
+. The directory `/etc/openstack`.
+
+| `clouds-file`
+| The file path of a clouds.yaml file. It is used if the `use-clouds` option is set to `true`.
+
+| `cloud`
+| The named cloud in the `clouds.yaml` file that you want to use. It is used if the `use-clouds` option is set to `true`.
+
+| `application-credential-id`
```
Comment (MaysaMacedo) [inline modules/cluster-cloud-controller-config-osp.adoc:97]: @imatza-rh these application-credentials options are being tested, correct?

Comment (imatza-rh) [reply modules/cluster-cloud-controller-config-osp.adoc:97]: @MaysaMacedo hey, the `application-credential-id` and `application-credential-secret` options have been tested

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 2 above)
Comment (rlobillo) [reply modules/cluster-cloud-controller-config-osp.adoc:257]: True, it was not tested. I think we should remove it from here.

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+| `application-credential-secret`
+| The secret of an application credential to authenticate with.
+
+| `tls-insecure`
+| Whether or not to verify the server's TLS certificate. If set to `true`, the certificate is not verified. By default, the certificate is verified.
+|===
+
+
+[id="ccm-config-networking-options"]
+== Networking options
```
[the remark this answers is not in the document]
Comment (rlobillo) [reply modules/cluster-cloud-controller-config-osp.adoc:112]: That is correct. Networking Options were not validated.

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 1 above)
Comment (rlobillo) [reply modules/cluster-cloud-controller-config-osp.adoc:193]: Right, and there are several restrictions on the LoadBalancer section that were filed as document bugs: 

https://issues.redhat.com/browse/OCPBUGS-2217
https://issues.redhat.com/browse/OCPBUGS-2350
https://issues.redhat.com/browse/OCPBUGS-2789

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+| `trust-id`
+| The Identity service trust ID. A trust represents the authorization of a user, or trustor, to delegate roles to another user, or trustee. Optionally, a trust authorizes the trustee to impersonate the trustor. You can find available trusts by querying the `/v3/OS-TRUST/trusts` endpoint of the Identity service API.
+
+| `trustee-id`
+| The Identity service trustee user ID.
+
+| `trustee-password`
+| The Identity service trustee user password.
+
+| `use-clouds`
```
Comment (rlobillo) [inline modules/cluster-cloud-controller-config-osp.adoc:80]: @MaysaMacedo We are always testing with below content:

    [Global]
    region      = regionOne
    ca-file     = /etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem
    use-clouds  = true
    clouds-file = /etc/openstack/secret/clouds.yaml
    cloud       = openstack

If I understood correctly, this configMap is not expected to be changed by the user, so... in which case we can have different configuration than above?

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+| `network-id`
+| The ID of the Networking service network on which load balancer VIPs are created. Unnecessary if `subnet-id` is set.
+
+// | `manage-security-groups`
+// | If the Neutron security groups should be managed separately. Default: false
+
+| `create-monitor`
+| Whether or not to create a health monitor for the service load balancer. A health monitor is required for services that declare `externalTrafficPolicy: Local`. The default value is `false`.
+
+This option is unsupported if you use the `ovn` provider.
```
Comment (dulek) [inline modules/cluster-cloud-controller-config-osp.adoc:195]: This is not true when using OSP17.

**Review by dulek — CHANGES_REQUESTED** [review]
Tiny remark.

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 6 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:195]:
```suggestion
This option is unsupported if you use {rh-openstack} earlier than version 17 and use the `ovn` provider.
```

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 5 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:80]: @MaysaMacedo Any thoughts on this?

**Comment by maxwelldb** [comment]
Added docs preview link.

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 6 above)
Comment (dulek) [reply modules/cluster-cloud-controller-config-osp.adoc:195]: Works for me!

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+// Commented options are untested as of yet.
+
+|===
+| Option | Description
+
+| `enabled`
+| Whether or not to enable the `LoadBalancer` type of services integration. The default value is `true`.
+
+| `use-octavia`
```
Comment (MaysaMacedo) [inline modules/cluster-cloud-controller-config-osp.adoc:152]: I would vote to remove this option from the docs or mention that use-octavia is always enforced as true because we don't support anything else than Octavia

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 1 above)
Comment (MaysaMacedo) [reply modules/cluster-cloud-controller-config-osp.adoc:193]: @maxwelldb will those restrictions be included in the release notes or in this docs?

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 5 above)
Comment (MaysaMacedo) [reply modules/cluster-cloud-controller-config-osp.adoc:80]: Since we're enforcing the UseClouds to true in the code, I believe there is no need to document the following options: auth-url, os-endpoint-type, username, password, domain related ones, tenant ones and user ones.

@stephenfin @dulek what do you think?

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+| Whether or not to use Octavia for the `LoadBalancer` type of service implementation rather than Neutron-LBaaS. The default value is `true`.
```
Comment (maxwelldb) [inline modules/cluster-cloud-controller-config-osp.adoc:153]:
```suggestion
// Removed: always enforced.
//| `use-octavia`
//| Whether or not to use Octavia for the `LoadBalancer` type of service implementation rather than Neutron-LBaaS. The default value is `true`.
```

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 1 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:194]: Likely both, as I read them now.

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 5 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:80]: @stephenfin @dulek ^^^

Comment (dulek) [reply modules/cluster-cloud-controller-config-osp.adoc:80]: I agree with Maysa here.

**Review by stephenfin — CHANGES_REQUESTED** [review]
Damn it, I had pending comments since *last* Monday. Sorry! :see_no_evil:

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 9]
```diff
@@ -18,12 +18,6 @@ The following options are used for {rh-openstack} CCM authentication with the {r
 |===
 | Option | Description
 
-| `auth-url`
-| The {rh-openstack} Identity service URL. For example, `http://128.110.154.166/identity`.
-
-| `os-endpoint-type`
-| Optional. The type of endpoint to use from the service catalog. If unset, public endpoints are used.
-
 | `ca-file`
```
Comment (stephenfin) [inline modules/cluster-cloud-controller-config-osp.adoc:21]: I actually think every one of these options in the general section is not used by us. They all look like things that would belong in a `cloud.yaml` file and therefore they are things the user should set there (since we're using `clouds.yaml` file), not here. @pierreprinetti @MaysaMacedo wdyt?

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 10]
```diff
@@ -0,0 +1,272 @@
+// Module included in the following assemblies:
```
[the remark this answers is not in the document]
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:1]: I'll attempt to add something that fits this description into #52520. This PR is more about adding the reference module.

[the remark this answers is not in the document]
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:1]: @stephenfin Thoughts on this? https://github.com/openshift/openshift-docs/pull/52520/commits/e8b486aa585800c8021fce6ee0ae2b6efc7a631a

**Comment by eurijon** [comment]
@maxwelldb hey, should this PR address the broken link in modules/installation-osp-setting-cloud-provider-options.adoc, or is it part of a different PR?
and also, if this PR applies to 4.12 only, should we re-open https://github.com/openshift/openshift-docs/pull/52294 to address the broken link issue in 4.9, 4.10 and 4.11?

Thanks!

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 9 above)
Comment (MaysaMacedo) [reply modules/cluster-cloud-controller-config-osp.adoc:21]: Good point, Stephen. I would say those are not needed given that we enforce to use cloud.yaml

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+| `floating-subnet`
+| Optional. A name pattern (glob or regular expression if starting with `~`) for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet`. If multiple subnets match the pattern, the first one with available IP addresses is used.
+
+| `floating-subnet-tags`
+| Optional. Tags for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet-tags`. If multiple subnets match these tags, the first one with available IP addresses is used.
+
+If the {rh-openstack} network is configured with sharing disabled, for example, with the `--no-share` flag used during creation, this option is unsupported. Set the network to share to use this option.
+
+| `lb-method`
+| The load balancing algorithm used to create the load balancer pool. The value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.
```
Comment (eurijon) [inline modules/cluster-cloud-controller-config-osp.adoc:116]: What about adding the provider here as well?
```suggestion
| The load balancing algorithm used to create the load balancer pool.
For the Amphora provider the value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.
```

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+
+For the OVN provider, only the `SOURCE_IP` algorithm is supported.
```
Comment (eurijon) [inline modules/cluster-cloud-controller-config-osp.adoc:118]: The algorithm for the OVN octavia provider is SOURCE_IP_PORT instead of SOURCE_IP.
For reference: https://access.redhat.com/documentation/fr-fr/red_hat_openstack_platform/16.2/html/using_octavia_for_load_balancing-as-a-service/create-other-kinds-lbs_rhosp-lbaas#create-ovn-lb_create-other-kinds-lbs
```suggestion
For the OVN provider, only the `SOURCE_IP_PORT` algorithm is supported.
```

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+
+For the Amphora provider, to use the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, you must configure the `"create-monitor=true` option in the `cloud-provider-config` namespace and `ETP:Local` on the load balancing service.
```
Comment (eurijon) [inline modules/cluster-cloud-controller-config-osp.adoc:120]: I think we can rephrase this to something like (to make it less mandatory):
```suggestion
For the Amphora provider, when using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, it's recommended to configure the `"create-monitor=true` option in the `cloud-provider-config` configMap on the `openshift-config` namespace and `ETP:Local` on the load balancer type service in order to get the balancing algorithm enforcement in the client to service endpoint connections.
```

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+
+| `lb-provider`
+| Optional. Used to specify the provider of the load balancer, for example, `amphora` or `octavia`. Only the Amphora and Octavia providers are supported.
+
+| `lb-version`
+| Optional. The load balancer API version. Only `"v2"` is supported.
+
+| `subnet-id`
+| The ID of the Networking servie subnet on which load balancer VIPs are created.
```
Comment (eurijon) [inline modules/cluster-cloud-controller-config-osp.adoc:129]: Typo
```suggestion
| The ID of the Networking service subnet on which load balancer VIPs are created. 
```

**Review by eurijon — COMMENTED** [review]
Added some remarks as well

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 15]
```diff
@@ -0,0 +1,296 @@
+// Module included in the following assemblies:
+//
+// * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc
+
+:_content-type: REFERENCE
+[id="cluster-cloud-controller-config_{context}"]
+= The {rh-openstack} Cloud Controller Manager (CCM) config file
+
+A {rh-openstack-first} CCM config file, called `cloud.conf` by convention, comprises several options in an INI format.
```
Comment (rlobillo) [inline modules/cluster-cloud-controller-config-osp.adoc:9]: @stephenfin @MaysaMacedo I think it is not clear, or at least for me, what config file is detailed here. AFAIK, user should modify only the configmap `cloud-provider-config` in the `openshift-config namespace`. Don't you think that should be mention at the top of the doc?

On the other hand, if the intention is to detail the content of the `cloud-conf` configMap in the `openshift-cloud-controller-manager`, all the parameters should be included in the informative section, as the user should never modify that configMap, right?

Comment (MaysaMacedo) [reply modules/cluster-cloud-controller-config-osp.adoc:9]: @rlobillo The config detailes here is the user managed config-map present on openshift-config namespace and it's ok if mention that here. I think we should document only the sections that will have an effect if modified by the user, which is what seems to be targeted in the PR

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 13 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:121]: @eurijon Is this the same idea? 

```suggestion
For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `"create-monitor=true` option in the `cloud-provider-config` configMap on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.
```

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 15 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc]: Check out the latest iteration of this PR and LMK.

Comment (rlobillo) [reply modules/cluster-cloud-controller-config-osp.adoc]: Hello @maxwelldb I think it's the other way around: The file that the user should modify is the configmap `cloud-provider-config` in the `openshift-config` namespace, and according to @MaysaMacedo and I agree with her, that should be one mentioned on the top of the doc.

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 16]
```diff
@@ -0,0 +1,324 @@
+// Module included in the following assemblies:
+//
+// * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc
+
+:_content-type: REFERENCE
+[id="cluster-cloud-controller-config_{context}"]
+= The {rh-openstack} Cloud Controller Manager (CCM) config map
+
+A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace. For example:
```
Comment (maxwelldb) [inline modules/cluster-cloud-controller-config-osp.adoc:9]:
```suggestion
A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `openshift-cloud-controller-manager` config map in the `cloud-conf` namespace. For example:
```

@rlobillo Is this what you indicated in your comment? 

If so, I need to swap out similar language in #52520. 

I want to be sure, as that other QEed PR and the example config map that I found both have `openshift-cloud-controller-manager` as the namespace. 

@MaysaMacedo @stephenfin

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 15 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc]: @rlobillo See https://github.com/openshift/openshift-docs/pull/53068/files#r1054406909

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 13 above)
Comment (eurijon) [reply modules/cluster-cloud-controller-config-osp.adoc:148]: @maxwelldb sure it is

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 16 above)
Comment (rlobillo) [reply modules/cluster-cloud-controller-config-osp.adoc:9]: @maxwelldb Nop, this change is not correct. Please discard it.

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 15 above)
Comment (rlobillo) [reply modules/cluster-cloud-controller-config-osp.adoc]: Hello Max! No, that is not what I meant, sorry, probably I wasn't clear. Let me rephrase starting from the beginning:

1. If the user wants to change something on CCM, it should modify below configMap (`cloud-provider-config` on `openshift-config` namespace):

```
$ oc get configMap/cloud-provider-config --namespace openshift-config -o json | jq -r .data.config 
[Global]
secret-name = openstack-credentials
secret-namespace = kube-system
region = regionOne
ca-file = /etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem
[LoadBalancer]
lb-provider = amphora
lb-method = SOURCE_IP
create-monitor = True
monitor-delay = 10s
monitor-timeout = 10s
monitor-max-retries = 1
floating-network-id = 03c0b9b2-5994-4611-a7dd-a95df903c2f9
```

2. Then, according to what it is explained on https://github.com/openshift/openshift-docs/pull/52520, the OCP cluster is internally (transparent for the user) converting that new configuration to the configMap called cloud-conf on the `openshift-cloud-controller-manager` namespace:

```
$ oc get configMap/cloud-conf --namespace openshift-cloud-controller-manager -o json | jq -r '.data["cloud.conf"]'
[Global]
region      = regionOne
ca-file     = /etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem
use-clouds  = true
clouds-file = /etc/openstack/secret/clouds.yaml
cloud       = openstack

[LoadBalancer]
lb-provider         = amphora
lb-method           = SOURCE_IP
create-monitor      = True
monitor-delay       = 10s
monitor-timeout     = 10s
monitor-max-retries = 1
floating-network-id = 03c0b9b2-5994-4611-a7dd-a95df903c2f9
use-octavia         = true
``` 
So, my point is that we should mention at the top of the doc that **if a user want to change something on the cluster, he/she should modify the cloud-provider-config configMap in the openshift-config namespace** and not in the configMap on the openshift-cloud-controller-manager namespace.

Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc]: :+1: How do you feel about this change, then: https://github.com/openshift/openshift-docs/pull/53068/commits/d73f8fc00b66ea729f698bbc55446c7d1a1fc52b

@rlobillo

Comment (rlobillo) [reply modules/cluster-cloud-controller-config-osp.adoc]: that is great, thanks!

**Comment by MaysaMacedo** [comment]
@maxwelldb looks good. Just one last question, do we plan to remove the following [configs](https://github.com/openshift/openshift-docs/pull/53068#discussion_r1039522553)?

**Comment by maxwelldb** [comment]
> @maxwelldb looks good. Just one last question, do we plan to remove the following [configs](https://github.com/openshift/openshift-docs/pull/53068#discussion_r1039522553)?

@MaysaMacedo I included those in the "these exist but you should not modify them" section at the bottom of the page based on one of @stephenfin's comments. If you think that they're unhelpful or misleading to include, though, I could pull them.

**Comment by maxwelldb** [comment]
@MaysaMacedo @stephenfin ^^^

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+// | `trust-id`
+// | The Identity service trust ID. A trust represents the authorization of a user, or trustor, to delegate roles to another user, or trustee. Optionally, a trust authorizes the trustee to impersonate the trustor. You can find available trusts by querying the `/v3/OS-TRUST/trusts` endpoint of the Identity service API.
+
+// | `trustee-id`
+// | The Identity service trustee user ID.
+
+// | `trustee-password`
+// | The Identity service trustee user password.
+
+// | `application-credential-id`
```
Comment (MaysaMacedo) [inline modules/cluster-cloud-controller-config-osp.adoc:75]: The options application-credential-id and application-credential-secret were tested by QE and the user is allowed to set it. Could we mention it somewhere?

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+// |===
+
+// ### Multi region support (alpha)
+
+// | environment variable `OS_CCM_REGIONAL` is set to `true` - allow CCM to set ProviderID with region name `${ProviderName}://${REGION}/${instance-id}`. Default: false.
+
+[id="cluster-cloud-controller-config-overrides"]
+== Options that the Operator overrides
+
+The CCM Operator overrides the following options, which you might recognize from configuring {rh-openstack}. Do not configure them yourself. They are included in this document for informational purposes only.
```
Comment (MaysaMacedo) [inline modules/cluster-cloud-controller-config-osp.adoc:259]: Looks ok to me

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 19]
```diff
@@ -0,0 +1,328 @@
+// Module included in the following assemblies:
+//
+// * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc
+
+:_content-type: REFERENCE
+[id="cluster-cloud-controller-config_{context}"]
+= The {rh-openstack} Cloud Controller Manager (CCM) config map
```
Comment (eurijon) [inline modules/cluster-cloud-controller-config-osp.adoc:7]: Should we remove RH reference from the CCM name as it's not provided/developed by Red Hat exclusively?
@MaysaMacedo @dulek @stephenfin please correct me if I'm wrong
```suggestion
= The OpenStack Cloud Controller Manager (CCM) config map
```

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 20]
```diff
@@ -0,0 +1,328 @@
...
+
+A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace.
```
Comment (eurijon) [inline modules/cluster-cloud-controller-config-osp.adoc:9]: Same here
```suggestion
A OpenStack CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace. 
```

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+
+A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace. 
+
+[IMPORTANT]
+====
+The `cloud-conf` config map is generated from the `cloud-provider-config` config map in the `openshift-config` namespace.
+
+To change the settings that are described by the `cloud-conf` config map, modify the `cloud-provider-config` config map. 
+
+As part of this synchronization, the CCM Operator overrides some options. For more information, see "The {rh-openstack} Cloud Controller Manager".
```
Comment (eurijon) [inline modules/cluster-cloud-controller-config-osp.adoc:17]: Where can we find the "The {rh-openstack} Cloud Controller Manager" reference? should it be a link?

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 19 above)
Comment (MaysaMacedo) [reply modules/cluster-cloud-controller-config-osp.adoc:7]: works for me to have only OpenStack

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 17 above)
Comment (stephenfin) [reply modules/cluster-cloud-controller-config-osp.adoc:75]: This is kind of by accident. The second we start setting this in `clouds.yaml` there will be a conflict. I'm not actually sure what happens in this situation but I would rather push people to configure stuff via `clouds.yaml` exclusively where possible.

```yaml
clouds:
  foo:
    auth:
      auth_url: <auth-url>
      application_credential_id: "<credential-id>"
      application_credential_secret: "<credential-secret>"
```

wdyt?

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 22]
```diff
@@ -4,53 +4,86 @@
...
-| Option | Description
+For example:
+
+.An example `cloud-conf` config map
+[source,yaml]
+----
+apiVersion: v1
+data:
+  cloud.conf: |
+    [Global] <1>
```
Comment (stephenfin) [inline modules/cluster-cloud-controller-config-osp.adoc:28]: ~Is `<1>` an asciidoc thing or a typo?~

Later: Ah, it's an asciidoc thing. I see the callout below.

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 19 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:7]: @stephenfin The idea is that I'm required to respect RH branding practices whenever I write about our products, so the PR is littered with "RHOSP" (`{rh-openstack}`). It's muscle memory. :) 

If OpenStack CCM is not subject to RH branding guidelines, I'm free to use "OpenStack" in place of "RHOSP."

My assumption here is that OpenStack CCM is usable outside of Red Hat's OpenStack solution. If that's not the case, let me know.

**Review by stephenfin — CHANGES_REQUESTED** [review]
Replied. This is looking better with just a few small comments outstanding.

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 21 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:17]: I have to merge #53068 prior to this PR going live. Link will be included in this document.

Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:17]: This comes from #52520. I can add a link after that's merged.

**Comment by eurijon** [comment]
lgtm

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 17 above)
Comment (MaysaMacedo) [reply modules/cluster-cloud-controller-config-osp.adoc:75]: ah ok, we can leave it out then

**Review by stephenfin — COMMENTED** [review]
/lgtm

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 23]
```diff
@@ -0,0 +1,328 @@
...
+
+If the {rh-openstack} network is configured with sharing disabled, for example, with the `--no-share` flag used during creation, this option is unsupported. Set the network to share to use this option.
+
+| `lb-method`
+| The load balancing algorithm used to create the load balancer pool.
+For the Amphora provider the value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.
+
+For the OVN provider, only the `SOURCE_IP_PORT` algorithm is supported.
+
+For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `"create-monitor=true` option in the `cloud-provider-config` configMap on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.
```
Comment (jeana-redhat) [inline modules/cluster-cloud-controller-config-osp.adoc:152]: Should there be a closing `"` in here?
```
`"create-monitor=true`
```

Comment (jeana-redhat) [inline modules/cluster-cloud-controller-config-osp.adoc:152]:
```suggestion
For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `"create-monitor=true` option in the `cloud-provider-config` config map on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.
```

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+ * `floating-subnet`
+ * `floating-subnet-tags`
+ * `network-id`
+ * `subnet-id`
+
+//  * `member-subnet-id`
+
+The behavior of these options is the same as that of the identically named options in the load balancer section of the CCM config file.
+
+You can set the `ClassName` value by specifying the Service annotation `loadbalancer.openstack.org/class`.
```
Comment (jeana-redhat) [inline modules/cluster-cloud-controller-config-osp.adoc:214]: Here and in line 236, should "Service" be lowercase?

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 25]
```diff
@@ -0,0 +1,328 @@
...
+// |===
+
+// ### Multi region support (alpha)
+
+// | environment variable `OS_CCM_REGIONAL` is set to `true` - allow CCM to set ProviderID with region name `${ProviderName}://${REGION}/${instance-id}`. Default: false.
+
+[id="cluster-cloud-controller-config-overrides"]
+== Options that the Operator overrides
+
+The CCM Operator overrides the following options, which you might recognize from configuring {rh-openstack}. Do not configure them yourself. They are included in this document for informational purposes only.
```
Comment (jeana-redhat) [inline modules/cluster-cloud-controller-config-osp.adoc:259]: Do you think it is worth making the "Do not..." into an admonition? Idk what the consequences are if they miss this.

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 26]
```diff
@@ -0,0 +1,328 @@
...
+CCM searches for the file in the following places:
+
+. The value of the `clouds-file` option.
+. A file path stored in the environment variable `OS_CLIENT_CONFIG_FILE`.
+. The directory `pkg/openstack`.
+. The directory `~/.config/openstack`.
+. The directory `/etc/openstack`.
+
+| `clouds-file`
+| The file path of a clouds.yaml file. It is used if the `use-clouds` option is set to `true`.
```
Comment (jeana-redhat) [inline modules/cluster-cloud-controller-config-osp.adoc:324]:
```suggestion
| The file path of a `clouds.yaml` file. It is used if the `use-clouds` option is set to `true`.
```

**Review by jeana-redhat — COMMENTED** [review]
Wheeee tables. Just a few small things from me, and a couple questions to evaluate. Otherwise LGTM!

/remove-label peer-review-in-progress
/remove-label peer-review-needed
/label peer-review-done

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 23 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:152]: Whoops. I'd actually change this to: 

```suggestion
For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `create-monitor` option as `true`  in the `cloud-provider-config` config map on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.
```

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 24 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:214]:
```suggestion
You can set the `ClassName` value by specifying the service annotation `loadbalancer.openstack.org/class`. 
```

File: modules/cluster-cloud-controller-config-osp.adoc [hunk 27]
```diff
@@ -0,0 +1,328 @@
...
+// | Reference to a tls container. This option works with Octavia, when this option is set then the cloud provider will create an Octavia Listener of type TERMINATED_HTTPS for a TLS Terminated loadbalancer.
+
+//  Format for tls container ref: `https://{keymanager_host}/v1/containers/{uuid}`
+//  Check `container-store` parameter if you want to disable validation.
+
+// | `container-store`
+// | Optional. Used to specify the store of the tls-container-ref, e.g. "barbican" or "external" - other store will cause a warning log. Default value - `barbican` - existence of tls container ref would always be performed. If set to `external` format for tls container ref will not be validated.
+
+| `max-shared-lb`
+| The maximum number of Services that can share a load balancer. The default value is `2`.
```
Comment (maxwelldb) [inline modules/cluster-cloud-controller-config-osp.adoc:236]:
```suggestion
| The maximum number of services that can share a load balancer. The default value is `2`.
```

File: modules/cluster-cloud-controller-config-osp.adoc (hunk 25 above)
Comment (maxwelldb) [reply modules/cluster-cloud-controller-config-osp.adoc:259]: Your effort is wasted--that's about it, AFAIK. I'm okay leaving as is, as the preceding sentence says as much.
